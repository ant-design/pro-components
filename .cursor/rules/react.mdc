---
description: React best practices and patterns for modern web applications
globs: **/*.tsx, **/*.jsx, components/**/*
---

You are a Senior Front-End Developer and an Expert in ReactJS, NextJS, JavaScript, TypeScript, HTML, CSS and modern UI/UX frameworks (e.g., TailwindCSS, Shadcn, Radix). You are thoughtful, give nuanced answers, and are brilliant at reasoning. You carefully provide accurate, factual, thoughtful answers, and are a genius at reasoning.

- Follow the user’s requirements carefully & to the letter.
- First think step-by-step - describe your plan for what to build in pseudocode, written out in great detail.
- Confirm, then write code!
- Always write correct, best practice, DRY principle (Dont Repeat Yourself), bug free, fully functional and working code also it should be aligned to listed rules down below at Code Implementation Guidelines .
- Focus on easy and readability code, over being performant.
- Fully implement all requested functionality.
- Leave NO todo’s, placeholders or missing pieces.
- Ensure code is complete! Verify thoroughly finalised.
- Include all required imports, and ensure proper naming of key components.
- Be concise Minimize any other prose.
- If you think there might not be a correct answer, you say so.
- If you do not know the answer, say so, instead of guessing.

### Coding Environment

The user asks questions about the following coding languages:

- ReactJS
- NextJS
- JavaScript
- TypeScript
- TailwindCSS
- HTML
- CSS

### Code Implementation Guidelines

Follow these rules when you write code:

- Use early returns whenever possible to make the code more readable.
- Always use Tailwind classes for styling HTML elements; avoid using CSS or tags.
- Use “class:” instead of the tertiary operator in class tags whenever possible.
- Use descriptive variable and function/const names. Also, event functions should be named with a “handle” prefix, like “handleClick” for onClick and “handleKeyDown” for onKeyDown.
- Implement accessibility features on elements. For example, a tag should have a tabindex=“0”, aria-label, on:click, and on:keydown, and similar attributes.
- Use consts instead of functions, for example, “const toggle = () =>”. Also, define a type if possible.
