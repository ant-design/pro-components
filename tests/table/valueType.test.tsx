import { ProProvider } from '@ant-design/pro-provider';
import ProTable from '@ant-design/pro-table';
import { cleanup, fireEvent, render } from '@testing-library/react';
import { Input } from 'antd';
import { act } from 'react';
import { waitForWaitTime } from '../util';

const cascaderOptions = [
  {
    field: 'front end',
    value: 'fe',
    language: [
      {
        field: 'Javascript',
        value: 'js',
      },
      {
        field: 'Typescript',
        value: 'ts',
      },
    ],
  },
  {
    field: 'back end',
    value: 'be',
    language: [
      {
        field: 'Java',
        value: 'java',
      },
      {
        field: 'Go',
        value: 'go',
      },
    ],
  },
];

const defaultProps = {
  columns: [
    {
      title: '标签',
      dataIndex: 'name',
      key: 'name',
      valueType: 'link',
      fieldProps: {
        color: 'red',
      },
    },
    {
      title: '树形下拉框',
      key: 'treeSelect',
      dataIndex: 'treeSelect',
      width: 100,
      fieldProps: {
        options: cascaderOptions,
        fieldNames: {
          children: 'language',
          label: 'field',
        },
        showSearch: true,
        filterTreeNode: true,
        multiple: true,
        treeNodeFilterProp: 'field',
      },
      valueType: 'treeSelect',
    },
  ],
  rowKey: 'key',
  request: () => {
    return Promise.resolve({
      total: 200,
      data: [
        {
          key: 0,
          name: 'TradeCode 0',
        },
      ],
      success: true,
    });
  },
};

afterEach(() => {
  cleanup();
});

describe('BasicTable valueType', () => {
  it('🎏 table support user valueType', async () => {
    const html = render(
      <ProProvider.Provider
        value={
          {
            valueTypeMap: {
              link: {
                render: (text: any) => <a id="link">{text}</a>,
                renderFormItem: (_: any, props: any) => (
                  <Input
                    placeholder="请输入链接"
                    id="name"
                    {...props?.fieldProps}
                  />
                ),
              },
            },
          } as any
        }
      >
        <ProTable
          form={{
            initialValues: { name: 'TradeCode' },
          }}
          {...defaultProps}
        />
      </ProProvider.Provider>,
    );
    await waitForWaitTime(1200);

    expect((await html.findAllByText('TradeCode 0')).length).toBe(1);

    expect(!!html.asFragment().querySelector('input#name')).toBeTruthy();

    expect(
      (html.asFragment().querySelector('input#name') as HTMLInputElement).value,
    ).toBe('TradeCode');

    html.unmount();
  });

  it('🎏 table valueType render support fieldProps', async () => {
    const html = render(
      <ProProvider.Provider
        value={
          {
            valueTypeMap: {
              link: {
                render: (text: any, { fieldProps }: any) => (
                  <a id="link">
                    {text}
                    {fieldProps.color}
                  </a>
                ),
                renderFormItem: (_: any, props: any) => (
                  <Input
                    placeholder="请输入链接"
                    id="name"
                    {...props?.fieldProps}
                  />
                ),
              },
            },
          } as any
        }
      >
        <ProTable
          form={{
            initialValues: { name: 'TradeCode' },
          }}
          {...defaultProps}
        />
      </ProProvider.Provider>,
    );
    await waitForWaitTime(1200);

    expect((await html.findAllByText('TradeCode 0red')).length).toBe(1);

    expect(!!html.asFragment().querySelector('input#name')).toBeTruthy();

    expect(
      (html.asFragment().querySelector('input#name') as HTMLInputElement).value,
    ).toBe('TradeCode');

    html.unmount();
  });
  it('🎏 table support filter when valueType is treeSelect', async () => {
    const html = render(<ProTable {...defaultProps} />);
    await waitForWaitTime(1200);

    act(() => {
      fireEvent.change(html.baseElement.querySelector('input#treeSelect')!, {
        target: {
          value: 'Ja',
        },
      });
    });
    await waitForWaitTime(300);
    expect(
      html.baseElement.querySelectorAll('span[title="Javascript"]').length,
    ).toBe(1);
    expect(html.baseElement.querySelectorAll('span[title="Java"]').length).toBe(
      1,
    );
    expect(
      html.baseElement.querySelectorAll('span[title="Typescript"]').length,
    ).toBe(0);
    expect(html.baseElement.querySelectorAll('span[title="Go"]').length).toBe(
      0,
    );

    act(() => {
      fireEvent.change(html.baseElement.querySelector('input#treeSelect')!, {
        target: {
          value: 'Javasc',
        },
      });
    });
    await waitForWaitTime(300);
    expect(
      html.baseElement.querySelectorAll('span[title="Javascript"]').length,
    ).toBe(1);
    expect(html.baseElement.querySelectorAll('span[title="Java"]').length).toBe(
      0,
    );
    expect(
      html.baseElement.querySelectorAll('span[title="Typescript"]').length,
    ).toBe(0);
    expect(html.baseElement.querySelectorAll('span[title="Go"]').length).toBe(
      0,
    );
    expect(html.asFragment()).toMatchSnapshot();

    html.unmount();
  });
});
