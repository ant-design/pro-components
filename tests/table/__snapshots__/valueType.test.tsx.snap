// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BasicTable valueType > üéè table support filter when valueType is treeSelect 1`] = `
<DocumentFragment>
  <div
    class="ant-pro-table"
  >
    <div
      class="ant-pro-card ant-pro-table-search ant-pro-table-search-query-filter"
    >
      <form
        autocomplete="off"
        class="ant-form ant-form-horizontal ant-pro-query-filter ant-pro-form"
      >
        <input
          style="display: none;"
          type="text"
        />
        <div
          class="ant-row ant-row-start ant-pro-query-filter-row"
          style="margin-left: -12px; margin-right: -12px;"
        >
          <div
            class="ant-col ant-col-8 ant-pro-query-filter-row-split"
            style="padding-left: 12px; padding-right: 12px;"
          >
            <div
              class="ant-form-item"
              style="flex-wrap: nowrap;"
            >
              <div
                class="ant-row ant-form-item-row"
              >
                <div
                  class="ant-col ant-form-item-label"
                  style="flex: 0 0 80px;"
                >
                  <label
                    class=""
                    for="name"
                    title="Ê†áÁ≠æ"
                  >
                    Ê†áÁ≠æ
                  </label>
                </div>
                <div
                  class="ant-col ant-form-item-control"
                  style="max-width: calc(100% - 80px);"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <span
                        class="ant-input-affix-wrapper ant-input-outlined"
                        style="width: 100%;"
                      >
                        <input
                          class="ant-input"
                          color="red"
                          id="name"
                          placeholder="ËØ∑ËæìÂÖ•"
                          type="text"
                          value=""
                        />
                        <span
                          class="ant-input-suffix"
                        >
                          <span
                            class="ant-input-clear-icon ant-input-clear-icon-hidden"
                            role="button"
                            tabindex="-1"
                          >
                            <span
                              aria-label="close-circle"
                              class="anticon anticon-close-circle"
                              role="img"
                            >
                              <svg
                                aria-hidden="true"
                                data-icon="close-circle"
                                fill="currentColor"
                                fill-rule="evenodd"
                                focusable="false"
                                height="1em"
                                viewBox="64 64 896 896"
                                width="1em"
                              >
                                <path
                                  d="M512 64c247.4 0 448 200.6 448 448S759.4 960 512 960 64 759.4 64 512 264.6 64 512 64zm127.98 274.82h-.04l-.08.06L512 466.75 384.14 338.88c-.04-.05-.06-.06-.08-.06a.12.12 0 00-.07 0c-.03 0-.05.01-.09.05l-45.02 45.02a.2.2 0 00-.05.09.12.12 0 000 .07v.02a.27.27 0 00.06.06L466.75 512 338.88 639.86c-.05.04-.06.06-.06.08a.12.12 0 000 .07c0 .03.01.05.05.09l45.02 45.02a.2.2 0 00.09.05.12.12 0 00.07 0c.02 0 .04-.01.08-.05L512 557.25l127.86 127.87c.04.04.06.05.08.05a.12.12 0 00.07 0c.03 0 .05-.01.09-.05l45.02-45.02a.2.2 0 00.05-.09.12.12 0 000-.07v-.02a.27.27 0 00-.05-.06L557.25 512l127.87-127.86c.04-.04.05-.06.05-.08a.12.12 0 000-.07c0-.03-.01-.05-.05-.09l-45.02-45.02a.2.2 0 00-.09-.05.12.12 0 00-.07 0z"
                                />
                              </svg>
                            </span>
                          </span>
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-col ant-col-8 ant-pro-query-filter-row-split"
            style="padding-left: 12px; padding-right: 12px;"
          >
            <div
              class="ant-form-item"
              style="flex-wrap: nowrap;"
            >
              <div
                class="ant-row ant-form-item-row"
              >
                <div
                  class="ant-col ant-form-item-label"
                  style="flex: 0 0 80px;"
                >
                  <label
                    class=""
                    for="treeSelect"
                    title="Ê†ëÂΩ¢‰∏ãÊãâÊ°Ü"
                  >
                    Ê†ëÂΩ¢‰∏ãÊãâÊ°Ü
                  </label>
                </div>
                <div
                  class="ant-col ant-form-item-control"
                  style="max-width: calc(100% - 80px);"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <div
                        class="ant-spin-nested-loading"
                      >
                        <div
                          class="ant-spin-container"
                        >
                          <div
                            class="ant-select ant-tree-select ant-select-outlined ant-select-in-form-item ant-pro-field-tree-select ant-select-multiple ant-select-allow-clear ant-select-show-arrow ant-select-open ant-select-show-search"
                            options="[object Object],[object Object]"
                            style="min-width: 60px; width: 100%;"
                          >
                            <div
                              class="ant-select-selector"
                            >
                              <span
                                class="ant-select-selection-wrap"
                              >
                                <div
                                  class="ant-select-selection-overflow"
                                >
                                  <div
                                    class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                                    style="opacity: 1;"
                                  >
                                    <div
                                      class="ant-select-selection-search"
                                      style="width: 0px;"
                                    >
                                      <input
                                        aria-autocomplete="list"
                                        aria-controls="treeSelect_list"
                                        aria-expanded="true"
                                        aria-haspopup="listbox"
                                        aria-owns="treeSelect_list"
                                        autocomplete="off"
                                        class="ant-select-selection-search-input"
                                        id="treeSelect"
                                        role="combobox"
                                        style=""
                                        type="search"
                                        value="Javasc"
                                      />
                                      <span
                                        aria-hidden="true"
                                        class="ant-select-selection-search-mirror"
                                      >
                                        Javasc¬†
                                      </span>
                                    </div>
                                  </div>
                                </div>
                              </span>
                            </div>
                            <span
                              aria-hidden="true"
                              class="ant-select-arrow"
                              style="user-select: none;"
                              unselectable="on"
                            >
                              <span
                                aria-label="search"
                                class="anticon anticon-search ant-select-suffix"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="search"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                                  />
                                </svg>
                              </span>
                            </span>
                            <span
                              aria-hidden="true"
                              class="ant-select-clear"
                              style="user-select: none;"
                              unselectable="on"
                            >
                              <span
                                aria-label="close-circle"
                                class="anticon anticon-close-circle"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="close-circle"
                                  fill="currentColor"
                                  fill-rule="evenodd"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M512 64c247.4 0 448 200.6 448 448S759.4 960 512 960 64 759.4 64 512 264.6 64 512 64zm127.98 274.82h-.04l-.08.06L512 466.75 384.14 338.88c-.04-.05-.06-.06-.08-.06a.12.12 0 00-.07 0c-.03 0-.05.01-.09.05l-45.02 45.02a.2.2 0 00-.05.09.12.12 0 000 .07v.02a.27.27 0 00.06.06L466.75 512 338.88 639.86c-.05.04-.06.06-.06.08a.12.12 0 000 .07c0 .03.01.05.05.09l45.02 45.02a.2.2 0 00.09.05.12.12 0 00.07 0c.02 0 .04-.01.08-.05L512 557.25l127.86 127.87c.04.04.06.05.08.05a.12.12 0 00.07 0c.03 0 .05-.01.09-.05l45.02-45.02a.2.2 0 00.05-.09.12.12 0 000-.07v-.02a.27.27 0 00-.05-.06L557.25 512l127.87-127.86c.04-.04.05-.06.05-.08a.12.12 0 000-.07c0-.03-.01-.05-.05-.09l-45.02-45.02a.2.2 0 00-.09-.05.12.12 0 00-.07 0z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="ant-col ant-col-8"
            style="padding-left: 12px; padding-right: 12px; text-align: end;"
          >
            <div
              class="ant-form-item ant-pro-query-filter-actions"
            >
              <div
                class="ant-row ant-form-item-row"
              >
                <div
                  class="ant-col ant-form-item-label"
                >
                  <label
                    class="ant-form-item-no-colon"
                    title=" "
                  >
                     
                  </label>
                </div>
                <div
                  class="ant-col ant-form-item-control"
                >
                  <div
                    class="ant-form-item-control-input"
                  >
                    <div
                      class="ant-form-item-control-input-content"
                    >
                      <div
                        class="ant-space ant-space-horizontal ant-space-align-center"
                        style="column-gap: 16px; row-gap: 16px;"
                      >
                        <div
                          class="ant-space-item"
                        >
                          <div
                            style="display: flex; gap: 8px; align-items: center;"
                          >
                            <button
                              class="ant-btn ant-btn-default ant-btn-color-default ant-btn-variant-outlined"
                              type="button"
                            >
                              <span>
                                Èáç ÁΩÆ
                              </span>
                            </button>
                            <button
                              class="ant-btn ant-btn-primary ant-btn-color-primary ant-btn-variant-solid"
                              type="button"
                            >
                              <span>
                                Êü• ËØ¢
                              </span>
                            </button>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
    <div
      class="ant-pro-card"
    >
      <div
        class="ant-pro-card-body"
        style="padding-block-start: 0;"
      >
        <div
          class="ant-pro-table-list-toolbar"
        >
          <div
            class="ant-pro-table-list-toolbar-container"
          >
            <div
              class="ant-pro-table-list-toolbar-left"
            />
            <div
              class="ant-pro-table-list-toolbar-right"
              style="align-items: center;"
            >
              <div
                class="ant-pro-table-list-toolbar-setting-items"
              >
                <div
                  class="ant-pro-table-list-toolbar-setting-item"
                >
                  <span>
                    <span
                      aria-label="reload"
                      class="anticon anticon-reload"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="reload"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M909.1 209.3l-56.4 44.1C775.8 155.1 656.2 92 521.9 92 290 92 102.3 279.5 102 511.5 101.7 743.7 289.8 932 521.9 932c181.3 0 335.8-115 394.6-276.1 1.5-4.2-.7-8.9-4.9-10.3l-56.7-19.5a8 8 0 00-10.1 4.8c-1.8 5-3.8 10-5.9 14.9-17.3 41-42.1 77.8-73.7 109.4A344.77 344.77 0 01655.9 829c-42.3 17.9-87.4 27-133.8 27-46.5 0-91.5-9.1-133.8-27A341.5 341.5 0 01279 755.2a342.16 342.16 0 01-73.7-109.4c-17.9-42.4-27-87.4-27-133.9s9.1-91.5 27-133.9c17.3-41 42.1-77.8 73.7-109.4 31.6-31.6 68.4-56.4 109.3-73.8 42.3-17.9 87.4-27 133.8-27 46.5 0 91.5 9.1 133.8 27a341.5 341.5 0 01109.3 73.8c9.9 9.9 19.2 20.4 27.8 31.4l-60.2 47a8 8 0 003 14.1l175.6 43c5 1.2 9.9-2.6 9.9-7.7l.8-180.9c-.1-6.6-7.8-10.3-13-6.2z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
                <div
                  class="ant-pro-table-list-toolbar-setting-item"
                >
                  <span>
                    <span
                      aria-label="column-height"
                      class="anticon anticon-column-height ant-dropdown-trigger"
                      role="img"
                      tabindex="-1"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="column-height"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M840 836H184c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h656c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8zm0-724H184c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h656c4.4 0 8-3.6 8-8v-60c0-4.4-3.6-8-8-8zM610.8 378c6 0 9.4-7 5.7-11.7L515.7 238.7a7.14 7.14 0 00-11.3 0L403.6 366.3a7.23 7.23 0 005.7 11.7H476v268h-62.8c-6 0-9.4 7-5.7 11.7l100.8 127.5c2.9 3.7 8.5 3.7 11.3 0l100.8-127.5c3.7-4.7.4-11.7-5.7-11.7H548V378h62.8z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
                <div
                  class="ant-pro-table-list-toolbar-setting-item"
                >
                  <span
                    aria-label="setting"
                    class="anticon anticon-setting"
                    role="img"
                    tabindex="-1"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="setting"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M924.8 625.7l-65.5-56c3.1-19 4.7-38.4 4.7-57.8s-1.6-38.8-4.7-57.8l65.5-56a32.03 32.03 0 009.3-35.2l-.9-2.6a443.74 443.74 0 00-79.7-137.9l-1.8-2.1a32.12 32.12 0 00-35.1-9.5l-81.3 28.9c-30-24.6-63.5-44-99.7-57.6l-15.7-85a32.05 32.05 0 00-25.8-25.7l-2.7-.5c-52.1-9.4-106.9-9.4-159 0l-2.7.5a32.05 32.05 0 00-25.8 25.7l-15.8 85.4a351.86 351.86 0 00-99 57.4l-81.9-29.1a32 32 0 00-35.1 9.5l-1.8 2.1a446.02 446.02 0 00-79.7 137.9l-.9 2.6c-4.5 12.5-.8 26.5 9.3 35.2l66.3 56.6c-3.1 18.8-4.6 38-4.6 57.1 0 19.2 1.5 38.4 4.6 57.1L99 625.5a32.03 32.03 0 00-9.3 35.2l.9 2.6c18.1 50.4 44.9 96.9 79.7 137.9l1.8 2.1a32.12 32.12 0 0035.1 9.5l81.9-29.1c29.8 24.5 63.1 43.9 99 57.4l15.8 85.4a32.05 32.05 0 0025.8 25.7l2.7.5a449.4 449.4 0 00159 0l2.7-.5a32.05 32.05 0 0025.8-25.7l15.7-85a350 350 0 0099.7-57.6l81.3 28.9a32 32 0 0035.1-9.5l1.8-2.1c34.8-41.1 61.6-87.5 79.7-137.9l.9-2.6c4.5-12.3.8-26.3-9.3-35zM788.3 465.9c2.5 15.1 3.8 30.6 3.8 46.1s-1.3 31-3.8 46.1l-6.6 40.1 74.7 63.9a370.03 370.03 0 01-42.6 73.6L721 702.8l-31.4 25.8c-23.9 19.6-50.5 35-79.3 45.8l-38.1 14.3-17.9 97a377.5 377.5 0 01-85 0l-17.9-97.2-37.8-14.5c-28.5-10.8-55-26.2-78.7-45.7l-31.4-25.9-93.4 33.2c-17-22.9-31.2-47.6-42.6-73.6l75.5-64.5-6.5-40c-2.4-14.9-3.7-30.3-3.7-45.5 0-15.3 1.2-30.6 3.7-45.5l6.5-40-75.5-64.5c11.3-26.1 25.6-50.7 42.6-73.6l93.4 33.2 31.4-25.9c23.7-19.5 50.2-34.9 78.7-45.7l37.9-14.3 17.9-97.2c28.1-3.2 56.8-3.2 85 0l17.9 97 38.1 14.3c28.7 10.8 55.4 26.2 79.3 45.8l31.4 25.8 92.8-32.9c17 22.9 31.2 47.6 42.6 73.6L781.8 426l6.5 39.9zM512 326c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm79.2 255.2A111.6 111.6 0 01512 614c-29.9 0-58-11.7-79.2-32.8A111.6 111.6 0 01400 502c0-29.9 11.7-58 32.8-79.2C454 401.6 482.1 390 512 390c29.9 0 58 11.6 79.2 32.8A111.6 111.6 0 01624 502c0 29.9-11.7 58-32.8 79.2z"
                      />
                    </svg>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-table-wrapper"
        >
          <div
            class="ant-spin-nested-loading"
          >
            <div
              class="ant-spin-container"
            >
              <div
                class="ant-table ant-table-middle"
              >
                <div
                  class="ant-table-container"
                >
                  <div
                    class="ant-table-content"
                  >
                    <table
                      style="table-layout: auto;"
                    >
                      <colgroup>
                        <col />
                        <col
                          style="width: 100px;"
                        />
                      </colgroup>
                      <thead
                        class="ant-table-thead"
                      >
                        <tr>
                          <th
                            class="ant-table-cell"
                            scope="col"
                          >
                            Ê†áÁ≠æ
                          </th>
                          <th
                            class="ant-table-cell"
                            scope="col"
                          >
                            Ê†ëÂΩ¢‰∏ãÊãâÊ°Ü
                          </th>
                        </tr>
                      </thead>
                      <tbody
                        class="ant-table-tbody"
                      >
                        <tr
                          class="ant-table-row ant-table-row-level-0"
                          data-row-key="0"
                        >
                          <td
                            class="ant-table-cell"
                          >
                            TradeCode 0
                          </td>
                          <td
                            class="ant-table-cell"
                          >
                            -
                          </td>
                        </tr>
                      </tbody>
                    </table>
                  </div>
                </div>
              </div>
              <ul
                class="ant-pagination ant-pagination-mini ant-table-pagination ant-table-pagination-right"
              >
                <li
                  class="ant-pagination-total-text"
                >
                  Á¨¨ 1-20 Êù°/ÊÄªÂÖ± 200 Êù°
                </li>
                <li
                  aria-disabled="true"
                  class="ant-pagination-prev ant-pagination-disabled"
                  title="‰∏ä‰∏ÄÈ°µ"
                >
                  <button
                    class="ant-pagination-item-link"
                    disabled=""
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      aria-label="left"
                      class="anticon anticon-left"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="left"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M724 218.3V141c0-6.7-7.7-10.4-12.9-6.3L260.3 486.8a31.86 31.86 0 000 50.3l450.8 352.1c5.3 4.1 12.9.4 12.9-6.3v-77.3c0-4.9-2.3-9.6-6.1-12.6l-360-281 360-281.1c3.8-3 6.1-7.7 6.1-12.6z"
                        />
                      </svg>
                    </span>
                  </button>
                </li>
                <li
                  class="ant-pagination-item ant-pagination-item-1 ant-pagination-item-active"
                  tabindex="0"
                  title="1"
                >
                  <a
                    rel="nofollow"
                  >
                    1
                  </a>
                </li>
                <li
                  class="ant-pagination-item ant-pagination-item-2"
                  tabindex="0"
                  title="2"
                >
                  <a
                    rel="nofollow"
                  >
                    2
                  </a>
                </li>
                <li
                  class="ant-pagination-item ant-pagination-item-3"
                  tabindex="0"
                  title="3"
                >
                  <a
                    rel="nofollow"
                  >
                    3
                  </a>
                </li>
                <li
                  class="ant-pagination-item ant-pagination-item-4"
                  tabindex="0"
                  title="4"
                >
                  <a
                    rel="nofollow"
                  >
                    4
                  </a>
                </li>
                <li
                  class="ant-pagination-item ant-pagination-item-5 ant-pagination-item-before-jump-next"
                  tabindex="0"
                  title="5"
                >
                  <a
                    rel="nofollow"
                  >
                    5
                  </a>
                </li>
                <li
                  class="ant-pagination-jump-next ant-pagination-jump-next-custom-icon"
                  tabindex="0"
                  title="ÂêëÂêé 5 È°µ"
                >
                  <a
                    class="ant-pagination-item-link"
                  >
                    <div
                      class="ant-pagination-item-container"
                    >
                      <span
                        aria-label="double-right"
                        class="anticon anticon-double-right ant-pagination-item-link-icon"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="double-right"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M533.2 492.3L277.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H188c-6.7 0-10.4 7.7-6.3 12.9L447.1 512 181.7 851.1A7.98 7.98 0 00188 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5zm304 0L581.9 166.1c-3-3.9-7.7-6.1-12.6-6.1H492c-6.7 0-10.4 7.7-6.3 12.9L751.1 512 485.7 851.1A7.98 7.98 0 00492 864h77.3c4.9 0 9.6-2.3 12.6-6.1l255.3-326.1c9.1-11.7 9.1-27.9 0-39.5z"
                          />
                        </svg>
                      </span>
                      <span
                        class="ant-pagination-item-ellipsis"
                      >
                        ‚Ä¢‚Ä¢‚Ä¢
                      </span>
                    </div>
                  </a>
                </li>
                <li
                  class="ant-pagination-item ant-pagination-item-10"
                  tabindex="0"
                  title="10"
                >
                  <a
                    rel="nofollow"
                  >
                    10
                  </a>
                </li>
                <li
                  aria-disabled="false"
                  class="ant-pagination-next"
                  tabindex="0"
                  title="‰∏ã‰∏ÄÈ°µ"
                >
                  <button
                    class="ant-pagination-item-link"
                    tabindex="-1"
                    type="button"
                  >
                    <span
                      aria-label="right"
                      class="anticon anticon-right"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="right"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M765.7 486.8L314.9 134.7A7.97 7.97 0 00302 141v77.3c0 4.9 2.3 9.6 6.1 12.6l360 281.1-360 281.1c-3.9 3-6.1 7.7-6.1 12.6V883c0 6.7 7.7 10.4 12.9 6.3l450.8-352.1a31.96 31.96 0 000-50.4z"
                        />
                      </svg>
                    </span>
                  </button>
                </li>
                <li
                  class="ant-pagination-options"
                >
                  <div
                    aria-label="È°µÁ†Å"
                    class="ant-select ant-select-sm ant-select-outlined ant-pagination-options-size-changer ant-select-single ant-select-show-arrow ant-select-show-search"
                  >
                    <div
                      class="ant-select-selector"
                    >
                      <span
                        class="ant-select-selection-wrap"
                      >
                        <span
                          class="ant-select-selection-search"
                        >
                          <input
                            aria-autocomplete="list"
                            aria-controls="rc_select_TEST_OR_SSR_list"
                            aria-expanded="false"
                            aria-haspopup="listbox"
                            aria-label="È°µÁ†Å"
                            aria-owns="rc_select_TEST_OR_SSR_list"
                            autocomplete="off"
                            class="ant-select-selection-search-input"
                            id="rc_select_TEST_OR_SSR"
                            role="combobox"
                            type="search"
                            value=""
                          />
                        </span>
                        <span
                          class="ant-select-selection-item"
                          title="20 Êù°/È°µ"
                        >
                          20 Êù°/È°µ
                        </span>
                      </span>
                    </div>
                    <span
                      aria-hidden="true"
                      class="ant-select-arrow"
                      style="user-select: none;"
                      unselectable="on"
                    >
                      <span
                        aria-label="down"
                        class="anticon anticon-down ant-select-suffix"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="down"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                          />
                        </svg>
                      </span>
                    </span>
                  </div>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
